### ISSUE TITLE
Flow doesn't refine arrays using `instanceof Array`

### ISSUE THREAD

**Issue by samwgoldman** [facebook/flow#472]
```
/* @flow */

function wrapArray(x: Array<number> | number): Array<number> {
  if (x instanceof Array) {
    return x;
  } else {
    return [x];
  }
}
```

Expected: `No errors!`

Actual:
```
test.js|3 col 23 error|  array type
|| This type is incompatible with
test.js|3 col 54 error|  number
|| 
```

Workaround: `Array.isArray` instead of `instanceof`

**Comment by impinball** [comment]
Here's the problem: this also happens `instanceof` check.

```js
var phony = Object.create(Array.prototype);
phony instanceof Array // true
Array.isArray(phony) // false
```

And this most certainly is not an Array instance. I doubt this is ever going to get fixed.

**Comment by mroch** [comment]
Closed by 05697a1.

I think `Object.create(Array.prototype) instanceof Array` already "worked" since `Object.create(Array.prototype)` creates an `InstanceT` of the `Array` class and `instanceof`'s job is to see if the RHS is in the prototype chain of the LHS, which it is in that case. On the other hand, `Array<T>` is special-cased to create an `ArrT`, and flow didn't know the relationship between `ArrT` and the `Array` class.

Fixing this also refines away other things that are clearly not possible, like `number` in the example above so I think it's worth doing.

**Comment by samwgoldman** [comment]
:clap:
